# <h1 style="text-align: center;">2. DESCRIPTIVE</h1>

## 1. How elderly is the patient population, and how common is underweight/malnutrition risk, especially among the oldest patients?
### Reasoning:  
Heart failure elderly patients often face malnutrition/cachexia, especially being underweight. So we first check: is the cohort elderly, and is underweight common.

### Metrics used:

Age category – shows risk across age groups

Gender – checks if risk differs by sex

BMI – basic nutrition screening

Underweight flag (BMI < 18.5) – clear risk indicator

### Why these:  
They’re simple, available for all patients, and good first-pass screens for malnutrition risk.

In [1]:
import pandas as pd

df_demography = pd.read_csv("Team05_PyCraft_Cleaned_data/demography_cleaned.csv")
age_order = ['21-29','29-39','39-49','49-59','59-69','69-79','79-89','89-110']

# ============================================================
# KPI GROUP 1: Population Overview
# ============================================================
total_patients = len(df_demography)
pct_female = round((df_demography['gender'] == 'Female').mean() * 100, 1)
pct_male = round((df_demography['gender'] == 'Male').mean() * 100, 1)
pct_elderly = round(df_demography['age_category'].isin(['69-79','79-89','89-110']).mean() * 100, 1)

print("=== KPI 1: POPULATION OVERVIEW ===")
print(f"Total patients:        {total_patients}")
print(f"% Female:              {pct_female}%")
print(f"% Male:                {pct_male}%")
print(f"% Elderly (69+):       {pct_elderly}%")
print()

# ============================================================
# KPI GROUP 2: Age & Gender Distribution (detail table)
# ============================================================
print("=== KPI 2: AGE & GENDER BREAKDOWN ===")
age_gender_table = pd.crosstab(df_demography['age_category'], df_demography['gender']).reindex(age_order)
print(age_gender_table)
print()

# ============================================================
# KPI GROUP 3: Nutritional Risk (BMI-based)
# ============================================================
df_demography['underweight'] = df_demography['bmi'] < 18.5

avg_bmi = round(df_demography['bmi'].mean(), 1)
pct_underweight_overall = round(df_demography['underweight'].mean() * 100, 1)
pct_underweight_oldest = round(
    df_demography[df_demography['age_category'] == '89-110']['underweight'].mean() * 100, 1
)

print("=== KPI 3: NUTRITIONAL RISK (BMI) ===")
print(f"Average BMI (overall):           {avg_bmi} kg/m²")
print(f"% Underweight (overall):         {pct_underweight_overall}%")
print(f"% Underweight (age 89-110):      {pct_underweight_oldest}%")
print()

print("BMI trend by age category:")
print(df_demography.groupby('age_category')['bmi'].mean().reindex(age_order).round(2))
print()

print("% Underweight by age category:")
print((df_demography.groupby('age_category')['underweight'].mean() * 100).reindex(age_order).round(1))

=== KPI 1: POPULATION OVERVIEW ===
Total patients:        2008
% Female:              57.9%
% Male:                42.1%
% Elderly (69+):       72.8%

=== KPI 2: AGE & GENDER BREAKDOWN ===
gender        Female  Male
age_category              
21-29              3     1
29-39              5     7
39-49             18    38
49-59             51    55
59-69            190   178
69-79            421   294
79-89            409   237
89-110            66    35

=== KPI 3: NUTRITIONAL RISK (BMI) ===
Average BMI (overall):           21.3 kg/m²
% Underweight (overall):         25.2%
% Underweight (age 89-110):      36.6%

BMI trend by age category:
age_category
21-29     24.92
29-39     22.41
39-49     22.17
49-59     21.78
59-69     21.48
69-79     21.23
79-89     21.14
89-110    20.85
Name: bmi, dtype: float64

% Underweight by age category:
age_category
21-29      0.0
29-39     25.0
39-49     17.9
49-59     22.6
59-69     22.8
69-79     25.3
79-89     26.0
89-110    36.6
Name: underweight, d

In [ ]:
import matplotlib.pyplot as plt

age_order = ['21-29','29-39','39-49','49-59','59-69','69-79','79-89','89-110']

# --- Chart 1: Age & Gender distribution with labels ---
age_gender_table = pd.crosstab(df_demography['age_category'], df_demography['gender']).reindex(age_order)

fig, ax = plt.subplots(figsize=(10, 5))
age_gender_table.plot(kind='bar', ax=ax)

for container in ax.containers:
    ax.bar_label(container, fontsize=8)

plt.title("Age & Gender Distribution")
plt.xlabel("Age Category")
plt.ylabel("Number of Patients")
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

# --- Chart 2: % Underweight by age category with labels ---
underweight_by_age = (df_demography.groupby('age_category')['underweight'].mean() * 100).reindex(age_order)

fig, ax = plt.subplots(figsize=(10, 5))
bars = ax.bar(underweight_by_age.index, underweight_by_age.values, color='orange')

ax.bar_label(bars, fmt='%.1f%%', fontsize=9)

plt.title("% of Patients Underweight (BMI < 18.5) by Age Category")
plt.xlabel("Age Category")
plt.ylabel("% Underweight")
plt.xticks(rotation=45)
plt.tight_layout()
plt.show()

### Insights:

72.8% are 69+ → truly elderly cohort

25.2% underweight overall

36.6% underweight in 89+ → strong age-linked trend

### How it helps:

Shows underweight is a major subgroup, not a minor issue

Justifies using BMI/underweight in predictive models

Supports targeted interventions (e.g., nutrition screening for 79+)

Confirms data cleaning worked (smooth age trend)

## 2. What proportion of patients died during hospitalization vs. were discharged home, and how does that break down by admission ward (ICU, Cardiology, General)? (hospitalization)

### Reasoning

Mortality is the clearest outcome measure — comparing it by ward shows whether sicker patients are actually being placed where they'd get closest monitoring.

What we measured

hospitalization_outcome (Alive/Dead/DischargeAgainstOrder) × admission_ward

In [ ]:
import pandas as pd

df_hospitalization = pd.read_csv("Team05_PyCraft_Cleaned_data/hospitalization_discharge_cleaned.csv")

# ============================================================
# KPI GROUP 1: Overall Discharge Outcome
# ============================================================
total_patients = len(df_hospitalization)
pct_died = round((df_hospitalization['hospitalization_outcome'] == 'Dead').mean() * 100, 2)
pct_alive = round((df_hospitalization['hospitalization_outcome'] == 'Alive').mean() * 100, 2)
pct_against_order = round((df_hospitalization['hospitalization_outcome'] == 'DischargeAgainstOrder').mean() * 100, 2)

print("=== KPI 1: OVERALL DISCHARGE OUTCOME ===")
print(f"Total patients:              {total_patients}")
print(f"% Died in-hospital:          {pct_died}%")
print(f"% Discharged alive:          {pct_alive}%")
print(f"% Discharged against order:  {pct_against_order}%")
print()

# ============================================================
# KPI GROUP 2: Outcome by Admission Ward
# ============================================================
print("=== KPI 2: OUTCOME BY ADMISSION WARD ===")
ward_counts = df_hospitalization['admission_ward'].value_counts()
print("Patients per ward:")
print(ward_counts)
print()

outcome_by_ward = (
    pd.crosstab(df_hospitalization['admission_ward'], df_hospitalization['hospitalization_outcome'], normalize='index') * 100
).round(1)
print("Outcome % by ward:")
print(outcome_by_ward)

In [ ]:
import matplotlib.pyplot as plt

fig, ax = plt.subplots(figsize=(10, 5))
outcome_by_ward.plot(kind='bar', ax=ax)

for container in ax.containers:
    ax.bar_label(container, fmt='%.1f%%', fontsize=8)

plt.title("Discharge Outcome by Admission Ward")
plt.xlabel("Admission Ward")
plt.ylabel("% of Patients")
plt.xticks(rotation=0)
plt.legend(title="Outcome")
plt.tight_layout()
plt.show()

## Insights

Key numbers
Overall in-hospital mortality: 0.55%
ICU mortality: 6.7% (~12x the overall rate)
ICU "discharge against order": 13.3% (highest of all wards)

Insight

Mortality is heavily concentrated in ICU — confirms ward placement reflects real severity, and makes admission_ward a strong feature for later prediction. Caveat: ICU has only 15 patients, so 6.7% = just 1 death — report as "1 of 15," not as a precise rate.


## 3. Which heart failure type is most dangerous
Reasoning:  
Heart failure isn’t one single disease — left, right, and biventricular failure behave differently. Since your cohort is elderly and comorbidity‑heavy, it’s important to check whether the type of heart failure itself is an independent risk factor for death or readmission.

Metrics used:

Heart failure type (Left / Right / Both)

Mortality at 6 months

Readmission at 6 months

LVEF (heart pumping efficiency)

Why these:  
They let you check whether worse outcomes in “Both” are due to true disease mechanism or simply lower pumping efficiency.

In [ ]:
import pandas as pd

df_cardiac_complications = pd.read_csv("Team05_PyCraft_Cleaned_data/cardiac_cleaned.csv")
df_hospitalization = pd.read_csv("Team05_PyCraft_Cleaned_data/hospitalization_discharge_cleaned.csv")

merged = df_cardiac_complications.merge(
    df_hospitalization[['patient_id', 'death_6m', 'readmission_6m', 'hospitalization_outcome']],
    on='patient_id', how='left'
)

# ============================================================
# KPI: Risk by Heart Failure Type
# ============================================================
print("=== Patients per heart failure type ===")
print(merged['heart_failure_type'].value_counts())
print()

print("=== 6-month mortality by type ===")
mortality_by_type = (merged.groupby('heart_failure_type')['death_6m'].mean() * 100).round(2)
print(mortality_by_type)
print()

print("=== 6-month readmission by type ===")
readmission_by_type = (merged.groupby('heart_failure_type')['readmission_6m'].mean() * 100).round(2)
print(readmission_by_type)
print()

print("=== Average pumping efficiency (LVEF) by type ===")
print(merged.groupby('heart_failure_type')['heart_pumping_efficiency'].mean().round(1))

In [ ]:
import matplotlib.pyplot as plt

hf_types = merged['heart_failure_type'].dropna().unique().tolist()
mortality_vals = mortality_by_type.reindex(hf_types)
readmission_vals = readmission_by_type.reindex(hf_types)

x = range(len(hf_types))
width = 0.35

fig, ax = plt.subplots(figsize=(8, 5))
bars1 = ax.bar([i - width/2 for i in x], mortality_vals, width, label='6-Month Mortality %')
bars2 = ax.bar([i + width/2 for i in x], readmission_vals, width, label='6-Month Readmission %')

ax.bar_label(bars1, fmt='%.1f%%', fontsize=8)
ax.bar_label(bars2, fmt='%.1f%%', fontsize=8)

ax.set_xticks(x)
ax.set_xticklabels(hf_types)
ax.set_xlabel("Heart Failure Type")
ax.set_ylabel("Percent of Patients")
ax.set_title("Risk by Heart Failure Type")
ax.legend()
plt.tight_layout()
plt.show()

## Insights
Biventricular (“Both”) is the riskiest: highest mortality (3.11%), highest readmission (42.2%), and most common (74%).

Right‑sided shows 0% mortality but has only 51 patients, so the sample is too small to interpret.

LVEF is similar across all types (49–53%), meaning the higher risk in “Both” isn’t just due to worse pumping — the combined left+right involvement adds extra risk.

### How it helps analysis:

Identifies a high‑risk clinical subgroup, not just a demographic one.

Justifies using heart_failure_type as a predictive model feature.

Shows the risk isn’t explained by LVEF alone → the type itself matters.

Sets up next steps: evaluate whether biventricular patients receive appropriately intensive monitoring or treatment.

## 4. What proportion of patients experienced readmission and mortality within 28 days, 3 months, and 6 months of discharge?
### Reasoning:  

Hospital readmissions and post-discharge mortality are key quality-of-care metrics.

Patients  28 days,3 months, and 6 months readmission rates are standard quality metrics hospitals report (e.g., CMS 30-day readmission measures use a similar window), so these numbers let you compare against national or peer benchmarks.Tracking these three windows over time or across departments highlights whether early (28-day, often "preventable") readmissions vs. later ones (3-6 month, often disease-progression driven) are the bigger problem.

Reporting mortality across the same three windows lets us compare this cohort against national or peer benchmarks using a consistent, internationally recognized time frame, rather than an arbitrary cutoff

### Metrics used:
The raw dataset gives us two ways to measure them:

<h6>Flag columns</h6> (readmission_28d, readmission_3m, readmission_6m, death_28d, etc.) — pre-computed binary indicators
<h6>Time-to-event columns</h6> (readmission_time_days, death_time_days) — the actual number of days until the event occurred

We used second method to measure readmission rate and mortality rate for accuracy instead of flag columns.

admission_date – shows when the patient was admitted

readmission_time_days -the actual number of days to readmit hospital

death_time_days -the actual number of days until the death


In [ ]:
#Reading the file 
df_hospitalization = pd.read_csv("Team05_PyCraft_Cleaned_data/hospitalization_discharge_cleaned.csv")

In [ ]:
df_hospitalization['admission_date'] = pd.to_datetime(df_hospitalization['admission_date'])
   
# Sort by patient and admission chronological order
df_hospitalization = df_hospitalization.sort_values(by=['patient_id', 'admission_date'])

# ============================================================
# Patients Readmission rate in 28 days,3 months, 6 months
# ============================================================
valid_discharges = df_hospitalization[df_hospitalization['readmission_time_days'] >= 0]
total_readmitted_patients = valid_discharges['patient_id'].nunique()

total_patients = df_hospitalization['patient_id'].nunique()


# Calculate proportions
rate_28d = (valid_discharges['readmission_time_days'] <= 28).sum() / total_patients
rate_3m = (valid_discharges['readmission_time_days'] <= 90).sum() / total_patients
rate_6m = (valid_discharges['readmission_time_days'] <= 180).sum() / total_patients


print("=== Readmission rate===")
print(f"Total patients:{total_patients}")
print(f"Total readmitted patients:{total_readmitted_patients}")
print(f"28-Day Readmission Rate: {rate_28d:.2%}")
print(f"3-Month Readmission Rate: {rate_3m:.2%}")
print(f"6-Month Readmission Rate: {rate_6m:.2%}")
print()
# ============================================================
# Patients Mortality rate in 28 days,3 months, 6 months
# ============================================================
valid_deaths = df_hospitalization[df_hospitalization['death_time_days'] >= 0]
total_died_patients = valid_deaths['patient_id'].nunique()

total_patients = df_hospitalization['patient_id'].nunique()


# Calculate proportions
mortality_28d = (valid_deaths['death_time_days'] <= 28).sum() / total_patients
mortality_3m = (valid_deaths['death_time_days'] <= 90).sum() / total_patients
mortality_6m = (valid_deaths['death_time_days'] <= 180).sum() / total_patients

print("=== Mortaility Rate ===")
print(f"Total patients:{total_patients}")
print(f"Total died patients:{total_died_patients}")
print(f"28-Day Mortality Rate: {mortality_28d:.2%}")
print(f"3-Month Mortality Rate: {mortality_3m:.2%}")
print(f"6-Month Mortality Rate: {mortality_6m:.2%}")
print()

### Insights:

Total patients:   2008<br/>
Total readmitted patients: 901<br/>
Total died patients:  44<br/>

28-Day Readmission Rate: 6.97% ,
3-Month Readmission Rate: 24.80% ,
6-Month Readmission Rate: 34.96%

28-Day Mortality Rate: 1.84% ,
3-Month Mortality Rate: 2.04% ,
6-Month Mortality Rate: 2.04%


1. Readmission is a far bigger problem than mortality in this cohort</li>
2. The two outcomes grow at very different rates over time
3. Mortality risk resolves early; readmission risk does not
    
Mortality-focused interventions (early follow-up, close monitoring) should be front-loaded into the first 1–3 months.<br/>
Readmission-focused interventions need to stay active across the full 6-month horizon, since new risk keeps emerging even in months 4–6.

## 5. How severe is heart failure at admission across genders, based on heart_failure_severity and heart_damage_level?

### Reasoning:
Heart failure severity is recorded on two scales:

- **heart_failure_severity** (levels 2–4): how much heart failure limits daily activity. Level 2 = mild limitation, level 4 = symptoms even at rest.
- **heart_damage_level** (levels 1–4): physical signs of heart failure on examination. Level 1 = no signs, level 4 = shock (the heart cannot pump enough blood).

Our cohort is 58% female. If one sex arrives sicker, later results by gender must take that into account.

### Metrics used:

Share of each sex (%) at every level of heart_failure_severity and heart_damage_level

Share of each sex in the most severe level of each scale

### Visual:
Each bar is 100% of the patients of one sex on one scale. Colours run from green (mildest) to dark red (most severe), so more red at the right-hand end means that sex arrives sicker.

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
from matplotlib.patches import Patch

df_demography = pd.read_csv("Team05_PyCraft_Cleaned_data/demography_cleaned.csv")
df_cardiac_complications = pd.read_csv("Team05_PyCraft_Cleaned_data/cardiac_cleaned.csv")

severity = df_demography[['patient_id', 'gender']].merge(
    df_cardiac_complications[['patient_id', 'heart_failure_severity', 'heart_damage_level']],
    on='patient_id', how='inner'
)

# Share of each sex (%) at every level of each scale
scales = ['heart_failure_severity', 'heart_damage_level']
pct = {s: pd.crosstab(severity['gender'], severity[s], normalize='index') * 100 for s in scales}

colours = {'heart_failure_severity': {2: '#7BC47F', 3: '#F2C14E', 4: '#B3261E'},
           'heart_damage_level': {1: '#7BC47F', 2: '#F2C14E', 3: '#E8833A', 4: '#B3261E'}}
most_severe_text = {'heart_failure_severity': 'level 4 (symptoms at rest)',
                    'heart_damage_level': 'level 4 (shock)'}
bars = [('heart_failure_severity', 'Female', 3.55), ('heart_failure_severity', 'Male', 2.85),
        ('heart_damage_level', 'Female', 1.25), ('heart_damage_level', 'Male', 0.55)]

fig, ax = plt.subplots(figsize=(12, 6))
for scale, sex, y in bars:
    left = 0
    for level, share in pct[scale].loc[sex].items():
        colour = colours[scale][level]
        ax.barh(y, share, left=left, height=0.65, color=colour, edgecolor='white')
        if share >= 6:  # write the level and % inside the segment when it fits
            ax.text(left + share / 2, y, f"Level {level}\n{share:.1f}%", ha='center', va='center',
                    fontsize=8.5, color='white' if colour == '#B3261E' else 'black')
        left += share
    worst = pct[scale].loc[sex].iloc[-1]
    ax.text(101.5, y, f"Most severe, {most_severe_text[scale]}: {worst:.1f}%",
            va='center', fontsize=9, fontweight='bold', color='#B3261E')

ax.text(0, 4.05, "heart_failure_severity  (2 = mild limitation  ...  4 = symptoms at rest)", fontsize=10, fontweight='bold', va='center')
ax.text(0, 1.75, "heart_damage_level  (1 = no signs  ...  4 = shock)", fontsize=10, fontweight='bold', va='center')
ax.axhline(2.15, color='grey', lw=0.8, ls='--')
ax.set_yticks([y for _, _, y in bars])
ax.set_yticklabels([sex for _, sex, _ in bars], fontsize=10)
ax.set_xlim(0, 100)
ax.set_ylim(0.1, 4.3)
ax.set_xlabel("% of patients within each sex (each bar = 100% of that sex)")
ax.set_title("Heart failure severity at admission by gender\n"
             "Each bar = all patients of one sex; each coloured block = share of them at that level",
             fontsize=12, pad=12)
ax.legend(handles=[Patch(color='#7BC47F', label='Mildest level'), Patch(color='#F2C14E', label='Moderate'),
                   Patch(color='#E8833A', label='Severe (heart_damage_level 3)'), Patch(color='#B3261E', label='Most severe level')],
          loc='upper center', bbox_to_anchor=(0.5, -0.12), ncol=4, fontsize=8.5, frameon=False)
for side in ['top', 'right']:
    ax.spines[side].set_visible(False)
plt.tight_layout()
plt.show()

## Insights

**Most patients of both sexes arrive with advanced heart failure.** About 81% of women and 85% of men are at heart_failure_severity level 3 or 4. The most common picture for both is heart_failure_severity level 3 with heart_damage_level 2.

**Men arrive slightly sicker on both scales:**
- **heart_failure_severity level 4 (symptoms at rest):** 34.0% of men vs 28.3% of women.
- **heart_damage_level level 4 (shock):** 4.0% of men vs 2.2% of women, almost double.
- **Mildest levels:** women are more often at the mildest level of both scales.

**The gap is small.** Level 3 of heart_damage_level is almost the same for both sexes (about 19–20%), so the extra severity in men is concentrated in shock. Shock is rare (60 patients), so that difference rests on small numbers.

### How it helps:

If men later show worse outcomes, part of the reason may simply be that they arrived sicker.

Supports using heart_failure_severity and heart_damage_level alongside gender in predictive models.

Flags shock (heart_damage_level 4) as the level to watch most closely at triage, especially in men.

## 6. How does the risk of death change over the first 4 weeks, based on the patient's consciousness level at admission?

### Reasoning:
Reduced consciousness in heart failure usually means the brain is not getting enough blood or oxygen, which is a warning sign of death. Following deaths day by day shows how dangerous each consciousness level is and **how early** that danger strikes.

`consciousness` levels, from least to most severe: Clear → ResponsiveToSound → ResponsiveToPain → Nonresponsive.

### Metrics used:

Consciousness level at admission (responsivenes_cleaned.csv)

death_time_days — day of death counted from admission (hospitalization_discharge_cleaned.csv)

Cumulative mortality (% of each group who have died) on every day up to day 28

### Visual:
Each line is one consciousness group and shows the % of that group who have died by each day. The higher and steeper the line, the more dangerous that level. Dots mark the end of each week.

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

df_responsivenes = pd.read_csv("Team05_PyCraft_Cleaned_data/responsivenes_cleaned.csv")
df_hospitalization = pd.read_csv("Team05_PyCraft_Cleaned_data/hospitalization_discharge_cleaned.csv")

mort = df_responsivenes[['patient_id', 'consciousness']].merge(
    df_hospitalization[['patient_id', 'death_time_days']], on='patient_id', how='inner'
)

consciousness_order = ['Clear', 'ResponsiveToSound', 'ResponsiveToPain', 'Nonresponsive']
colours = {'Clear': '#4C72B0', 'ResponsiveToSound': '#DD8452', 'ResponsiveToPain': '#C44E52', 'Nonresponsive': '#8C0000'}
days = np.arange(0, 29)

fig, ax = plt.subplots(figsize=(11, 6))
for level in consciousness_order:
    group = mort[mort['consciousness'] == level]
    # % of the group who have died by each day (patients with no death have no death_time_days)
    cumulative = [(group['death_time_days'] <= d).mean() * 100 for d in days]
    ax.step(days, cumulative, where='post', color=colours[level], lw=2.5, label=f"{level} ({len(group)} patients)")
    ax.plot([7, 14, 21, 28], [cumulative[d] for d in [7, 14, 21, 28]], 'o', color=colours[level], ms=6)
    ax.text(28.6, max(cumulative[28], 5), f"{cumulative[28]:.1f}% died\nby week 4", va='center', fontsize=9,
            fontweight='bold', color=colours[level])
    if level == 'Nonresponsive':  # show the weekly climb for the most severe group
        for d in [7, 14, 21]:
            ax.annotate(f"{cumulative[d]:.1f}%", (d, cumulative[d]), textcoords='offset points', xytext=(0, 9),
                        ha='center', fontsize=9, color=colours[level])

for d in [7, 14, 21, 28]:
    ax.axvline(d, color='grey', ls=':', lw=1)
ax.set_xticks([0, 7, 14, 21, 28])
ax.set_xticklabels(['Admission', 'End of week 1', 'End of week 2', 'End of week 3', 'End of week 4'])
ax.set_xlim(0, 32.5)
ax.set_ylim(0, 100)
ax.set_ylabel("% of patients in the group who have died")
ax.set_title("Mortality over the first 4 weeks by consciousness level at admission\n"
             "Each line = one consciousness group; height = % of that group who have died by that day",
             fontsize=12)
ax.legend(title="Consciousness at admission", loc='center right', bbox_to_anchor=(0.9, 0.45), fontsize=9)
for side in ['top', 'right']:
    ax.spines[side].set_visible(False)
plt.tight_layout()
plt.show()

## Insights

**The worse the consciousness, the higher the death rate.** By the end of week 4:
- Clear: 1.2% have died (23 of 1,974)
- ResponsiveToSound: 15.8% (3 of 19)
- ResponsiveToPain: 25.0% (1 of 4)
- Nonresponsive: 90.9% (10 of 11)

**Deaths happen early.** About two-thirds of all deaths in the first 4 weeks occur in week 1. More than half of Nonresponsive patients (54.5%) have died by the end of week 1. The other impaired groups have no further deaths after week 1, which is why their lines go flat.

**Rare but dangerous.** Only 1.7% of patients (34) arrive with reduced consciousness, yet they account for 38% of all deaths in the first 4 weeks.

**Caveat:** the groups with reduced consciousness are small (4–19 patients), so their exact percentages can move a lot with one death. The overall pattern is still clear.

### How it helps:

Shows that the first week is the critical window: patients with reduced consciousness need urgent review on admission.

Supports using consciousness level as a strong early-death predictor in the predictive models, as an ordered feature, since risk rises with each level.

Identifies Nonresponsive patients as a group where early conversations with families are appropriate.

## 7. Which medications are most commonly prescribed?

### Reasoning:

Understanding the most frequently prescribed medications helps identify standard treatment patterns for heart-failure patients and reveals which therapies are most commonly used in routine clinical practice. The prescription data contains all medications administered during patient hospitalization.

## Metric Used:
Prescription Frequency
Number of Times Each Drug Appears
Percentage of Patients Receiving Each Medication

## Why These:

Prescription frequency is a simple and effective measure for identifying the medications most commonly used by clinicians. Frequently prescribed drugs often represent guideline-based treatment and standard care practices.

In [ ]:
import pandas as pd
import numpy as np
df_cardiac = pd.read_csv("Team05_PyCraft_Cleaned_data/cardiac_cleaned.csv")
df_demography = pd.read_csv("Team05_PyCraft_Cleaned_data/demography_cleaned.csv")
df_hospitalization = pd.read_csv("Team05_PyCraft_Cleaned_data/hospitalization_discharge_cleaned.csv")
df_labs = pd.read_csv("Team05_PyCraft_Cleaned_data/labs_cleaned.csv")
df_patienthistory = pd.read_csv("Team05_PyCraft_Cleaned_data/patienthistory_cleaned.csv")
df_prescriptions = pd.read_csv("Team05_PyCraft_Cleaned_data/precriptions_cleaned.csv")
df_responsiveness = pd.read_csv("Team05_PyCraft_Cleaned_data/responsivenes_cleaned.csv")


In [ ]:
top_drugs = (
    df_prescriptions['drug_name']
    .value_counts()
    .reset_index()
)

top_drugs.columns = [
    'Medication',
    'Prescription_Count'
]

top_drugs.head(20)

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

top20 = (
    df_prescriptions['drug_name']
    .value_counts()
    .head(20)
    .reset_index()
)

top20.columns = [
    'Medication',
    'Prescription_Count'
]

plt.figure(figsize=(12,8))

sns.barplot(
    data=top20,
    x='Prescription_Count',
    y='Medication'
)

plt.title("Top 20 Most Frequently Prescribed Medications")
plt.xlabel("Number of Prescriptions")
plt.ylabel("Medication")

plt.show()

## Insights:

The most frequently appearing medications in the dataset include:

Spironolactone Tablet
Furosemide Tablet
Furosemide Injection
Digoxin Tablet
Meglumine Adenosine Cyclophosphate For Injection
Atorvastatin Calcium Tablet
Aspirin Enteric-Coated Tablet
Deslanoside Injection
Metoprolol Succinate Sustained-Release Tablet
Clopidogrel Hydrogen Sulphate Tablet

These medications indicate a strong focus on:

Fluid management (diuretics)
Heart-rate control
Cardiac function improvement
Blood-thinning and cardiovascular protection

## How it Helps:
Identifies the most commonly used treatment strategies.
Highlights adherence to heart-failure treatment guidelines.
Supports medication utilization analysis.
Helps evaluate which therapies may be associated with better outcomes in later predictive analysis.
Provides evidence for treatment optimization and clinical decision support.

## 8. Which Laboratory Markers Are Strongest Mortality Indicators?

## Reasoning:
Laboratory biomarkers provide an objective view of a patient's physiological condition and are often early indicators of disease severity. By comparing laboratory values between survivors and non-survivors (death_6m), we can identify biomarkers most strongly associated with mortality risk. labs_cleaned.csv and hospitalization_discharge_cleaned were used for this analysis.

## Metric Used:
6-Month Mortality (death_6m)
Mean Difference Between Survivors and Non-Survivors
Statistical Significance (p-value)
Random Forest Feature Importance (for ranking biomarkers)
## Why These:

These metrics help identify laboratory markers that show the greatest separation between low-risk and high-risk patients. Feature importance and statistical testing provide evidence of which biomarkers contribute most to mortality prediction.


In [ ]:
death_6m = 1   # Died within 6 months
death_6m = 0   # Survived

In [ ]:
lab_df = df_labs.merge(
    df_hospitalization[
        ['patient_id', 'death_6m']
    ],
    on='patient_id',
    how='inner'
)

In [ ]:
lab_cols = ['patient_id','temp','pulse','resp','sbp','dbp','MAP','FiO2','creatinine','urea',
 'uric_acid','glomerular_filtration_rate','cystatin','WBC','monocyte_ratio',
 'monocyte_count','RBC','coefficient_of_variation_RBC_DistributionWidth',
 'Standard_Deviation_RBC_DistributionWidth','MCV','HCT','lymphocyte_count','MCH','MCHC','MPV',
 'basophil_ratio','basophil_count','eosinophil_ratio','eosinophil_count','hemoglobin','platelet',
 'platelet_distribution_width','platelet_hematocrit','neutrophil_ratio','neutrophil_count',
 'd_dimer','international_normalized_ratio','activated_partial_thromboplastin_time',
 'thrombin_time','prothrombin_activity', 'prothrombin_time_ratio','fibrinogen',
 'high_sensitivity_troponin','myoglobin','CO2_capacity', 'calcium','potassium', 'chloride', 'sodium',
 'phosphorus','magnesium', 'creatine_kinase_isoenzyme_to_creatine_kinase',
 'hydroxybutyrate_dehydrogenase_to_lactate_dehydrogenase','hydroxybutyrate_dehydrogenase',
 'glutamic_oxaloacetic_transaminase', 'creatine_kinase', 'creatine_kinase_isoenzyme',
 'lactate_dehydrogenase', 'brain_natriuretic_peptide', 'high_sensitivity_protein', 'nucleotidase',
 'fucosidase', 'albumin', 'white_globulin_ratio', 'glutamyltranspeptidase',
 'glutamic_pyruvic_transaminase', 'glutamic_oxaliplatin', 'indirect_bilirubin',
 'ALP', 'globulin', 'direct_bilirubin', 'total_bilirubin', 'total_bile_acid',
 'total_protein', 'erythrocyte_sedimentation_rate','cholesterol', 'LDL_C', 'triglyceride',
 'HDL_C', 'homocysteine', 'apolipoprotein_a', 'apolipoprotein_b', 'lipoprotein', 'pH',
 'standard_residual_base', 'standard_bicarbonate', 'pCO2', 'tCO2','methemoglobin', 'HCT_BG',
 'HHb', 'K_ion', 'Cl_ion', 'Na_ion', 'glucose_blood_gas', 'lactate', 'measured_residual_base',
 'HCO3_act', 'CO_Hb', 'temp_BG', 'sO2', 'pO2', 'O2Hb', 'anion_gap', 'free_calcium', 'total_hemoglobin']


In [ ]:
mortality_lab = []

for col in lab_cols:

    survived_mean = (
        lab_df.loc[
            lab_df['death_6m'] == 0,
            col
        ].mean()
    )

    died_mean = (
        lab_df.loc[
            lab_df['death_6m'] == 1,
            col
        ].mean()
    )

    difference = abs(
        died_mean - survived_mean
    )

    mortality_lab.append([
        col,
        survived_mean,
        died_mean,
        difference
    ])

mortality_lab = pd.DataFrame(
    mortality_lab,
    columns=[
        'Lab_Marker',
        'Survived_Mean',
        'Died_Mean',
        'Difference'
    ]
)

mortality_lab.sort_values(
    'Difference',
    ascending=False
)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

X = (
    lab_df[lab_cols]
    .fillna(
        lab_df[lab_cols].median()
    )
)

y = lab_df['death_6m']

rf = RandomForestClassifier(
    n_estimators=200,
    random_state=42
)

rf.fit(X, y)

importance_df = pd.DataFrame({
    'Marker': lab_cols,
    'Importance': rf.feature_importances_
})

importance_df = (
    importance_df
    .sort_values(
        'Importance',
        ascending=False
    )
)

print(importance_df)

In [ ]:
# Keep only Top 20 markers

top20_markers = (
    importance_df
    .sort_values(
        'Importance',
        ascending=False
    )
    .head(20)
)

import seaborn as sns
import matplotlib.pyplot as plt

plt.figure(figsize=(12,8))

sns.barplot(
    data=top20_markers,
    x='Importance',
    y='Marker',
    
)

plt.title(
    'Top 20 Laboratory Markers Associated with 6-Month Mortality'
)

plt.xlabel('Feature Importance')
plt.ylabel('Laboratory Marker')

plt.tight_layout()

plt.show()

## Insights:
Elevated BNP levels typically indicate worsening heart failure and increased cardiac stress.
Higher Troponin levels suggest myocardial injury and are often linked to poorer outcomes.
Increased Creatinine reflects impaired kidney function, a major risk factor in heart-failure patients.
Lower Hemoglobin levels may indicate anemia, reducing oxygen delivery and worsening prognosis.
Patients with abnormal cardiac, renal, and hematologic biomarkers tend to experience higher mortality risk.
## How it Helps:
Identifies high-risk patients at admission.
Supports early clinical intervention.
Improves mortality prediction models.
Enables risk stratification and personalized treatment.
Helps clinicians focus on the most informative biomarkers when monitoring patient progression.

## 9. What proportion of patients have each comorbidity?

## Reasoning:

Comorbidities significantly influence heart-failure progression, mortality, and readmission risk. Understanding the prevalence of chronic conditions helps identify the most common health burdens affecting the patient population and provides insight into overall disease complexity. Data was obtained from patienthistory_cleaned.

## Metric Used:
Comorbidity Prevalence (%)
Percentage of patients with each disease:
CCI Score for overall disease burden.

## Why These:

Prevalence percentages reveal which chronic diseases are most frequently observed in the patient cohort. CCI score provides a standardized measure of cumulative comorbidity burden and is widely used for risk assessment.

In [ ]:
comorbidity_cols = [
    'cerebrovascular_disease',
    'dementia',
    'chronic_obstructive_pulmonary_disease',
    'connective_tissue_disease',
    'peptic_ulcer_disease',
    'diabetes',
    'moderate_to_severe_chronic_kidney_disease',
    'hemiplegia',
    'leukemia',
    'malignant_lymphoma',
    'solid_tumor',
    'liver_disease',
    'aids',
    'type2_respiratory_failure',
    'acute_renal_failure'
]

comorbidity_percent = (
    df_patienthistory[comorbidity_cols]
    .mean()
    .mul(100)
    .sort_values(ascending=False)
    .reset_index()
)

comorbidity_percent.columns = [
    'Comorbidity',
    'Patient_Percentage'
]

print(comorbidity_percent)

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

plt.figure(figsize=(10,6))

sns.barplot(
    data=comorbidity_percent,
    x='Patient_Percentage',
    y='Comorbidity'
)

plt.xlabel('Percentage of Patients (%)')
plt.ylabel('Comorbidity')
plt.title('Prevalence of Comorbidities')

plt.tight_layout()

plt.show()

## Insights:

The analysis focused on the following comorbidities:

Cerebrovascular Disease,
Dementia,
Chronic Obstructive Pulmonary Disease (COPD),
Connective Tissue Disease,
Peptic Ulcer Disease,
Diabetes,
Moderate to Severe Chronic Kidney Disease (CKD),
Hemiplegia,
Leukemia,
Malignant Lymphoma,
Solid Tumor,
Liver Disease,
AIDS,
Type 2 Respiratory Failure,
Acute Renal Failure.

Key observations include:

Diabetes, CKD, COPD, and cerebrovascular disease are often among the most prevalent chronic conditions in heart-failure populations.
Patients frequently present with multiple comorbidities rather than a single disease.
Higher comorbidity burden is associated with increased clinical complexity and poorer outcomes.

## How it Helps:
Identifies the most common chronic conditions in the population.
Supports risk stratification and patient profiling.
Helps clinicians prioritize screening and monitoring efforts.
Provides valuable features for mortality and readmission prediction models.
Informs healthcare resource allocation and disease-management strategies. 

## 10. Which comorbidity combination of patients have high risk?

### Reasoning: 
Patients with multiple chronic conditions generally have poorer outcomes than those with a single disease. We analyzed comorbidity burden and disease combinations to identify groups with the highest 6-month mortality and readmission risk using patienthistory_cleaned and hospitalization_discharge_cleaned.

### Metrix Used: 
6-Month Mortality Rate (death_6m)
6-Month Readmission Rate (readmission_6m)
Charlson Comorbidity Index (cci_score)
Comorbidity Count (total diseases per patient)

### Why These:
These metrics directly measure patient outcomes and are widely used to assess disease burden and long-term clinical risk. The CCI score summarizes the impact of multiple comorbidities into a single severity measure.


In [ ]:
risk_df = df_patienthistory.merge(
    df_hospitalization[
        ['patient_id',
         'death_6m',
         'readmission_6m']
    ],
    on='patient_id',
    how='left'
)

In [ ]:
comorbidity_cols = [
    'cerebrovascular_disease',
    'dementia',
    'chronic_obstructive_pulmonary_disease',
    'connective_tissue_disease',
    'peptic_ulcer_disease',
    'diabetes',
    'moderate_to_severe_chronic_kidney_disease',
    'hemiplegia',
    'leukemia',
    'malignant_lymphoma',
    'solid_tumor',
    'liver_disease',
    'aids',
    'type2_respiratory_failure',
    'acute_renal_failure'
]

In [ ]:
# create diesese combination
risk_df['disease_combination'] = (
    risk_df[comorbidity_cols]
    .apply(
        lambda row: ', '.join(
            [col for col in comorbidity_cols
             if row[col] == 1]
        ),
        axis=1
    )
)

risk_df['disease_combination'] = (
    risk_df['disease_combination']
    .replace('', 'No_Comorbidity')
)

In [ ]:
#highest mortality risk combination
combo_risk = (
    risk_df.groupby('disease_combination')
           .agg(
               Patients=('patient_id', 'count'),
               Deaths=('death_6m', 'sum')
           )
)

combo_risk['Mortality_Rate'] = (
    combo_risk['Deaths']
    /
    combo_risk['Patients']
    * 100
)

combo_risk = combo_risk[
    combo_risk['Patients'] >= 10
]

combo_risk = combo_risk.sort_values(
    'Mortality_Rate',
    ascending=False
)

combo_risk.head(20)

In [ ]:
#Highest readmission risk combination
combo_readmission = (
    risk_df.groupby('disease_combination')
           .agg(
               Patients=('patient_id', 'count'),
               Readmissions=('readmission_6m', 'sum')
           )
)

combo_readmission['Readmission_Rate'] = (
    combo_readmission['Readmissions']
    /
    combo_readmission['Patients']
    * 100
)

combo_readmission = combo_readmission[
    combo_readmission['Patients'] >= 10
]

combo_readmission = combo_readmission.sort_values(
    'Readmission_Rate',
    ascending=False
)

combo_readmission.head(20)

In [ ]:
risk_df['comorbidity_count'] = (
    risk_df[comorbidity_cols]
    .fillna(0)
    .sum(axis=1)
)

risk_df.groupby('comorbidity_count')['death_6m'].mean() * 100

### Insights:
Patients with higher comorbidity counts showed higher mortality and readmission rates.
Conditions such as moderate_to_severe_chronic_kidney_disease, acute_renal_failure, type2_respiratory_failure, and cerebrovascular_disease are expected to contribute substantially to risk when combined with heart failure.
Higher CCI scores indicate greater overall clinical complexity and poorer prognosis.

### How it Helps: 
Identifies high-risk patients early.
Supports targeted monitoring and follow-up.
Helps prioritize healthcare resources.
Improves mortality and readmission prediction models.
Enables personalized treatment planning for patients with multiple chronic conditions.